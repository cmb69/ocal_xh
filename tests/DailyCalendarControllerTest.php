<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Ocal_XH.
 *
 * Ocal_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ocal_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ocal_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Ocal;

use DateTimeImmutable;
use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use XH\CSRFProtection as CsrfProtector;

class DailyCalendarControllerTest extends TestCase
{
    /** @var DailyCalendarController */
    private $sut;

    /** @var ListService&MockObject */
    private $listService;

    public function setUp(): void
    {
        global $plugin_cf;

        $_SERVER['QUERY_STRING'] = "";
        $csrfProtector = $this->createStub(CsrfProtector::class);
        $csrfProtector->method('tokenInput')->willReturn(
            '<input type="hidden" name="xh_csrf_token" value="dcfff515ebf5bd421d5a0777afc6358b">'
        );
        $plugin_cf = XH_includeVar("./config/config.php", 'plugin_cf');
        $config = $plugin_cf['ocal'];
        $plugin_tx = XH_includeVar("./languages/en.php", 'plugin_tx');
        $lang = $plugin_tx['ocal'];
        $now = new DateTimeImmutable("2023-07-02");
        $this->listService = $this->createStub(ListService::class);
        $db = $this->createStub(Db::class);
        $db->method('findOccupancy')->willReturn(new DailyOccupancy("test-daily", 3));
        $this->sut = new DailyCalendarController(
            "/",
            "./",
            $csrfProtector,
            $config,
            $lang,
            $now,
            $this->listService,
            $db,
            true,
            "test-daily",
            1
        );
    }

    public function testDefaultActionRendersCalendar(): void
    {
        $response = $this->sut->defaultAction("test-daily", 1);
        Approvals::verifyHtml($response->output());
    }

    public function testDefaultActionHandlesAjaxRequest(): void
    {
        $_SERVER['HTTP_X_REQUESTED_WITH'] = 'XMLHttpRequest';
        $_GET = ['ocal_name' => "test-daily"];
        $response = $this->sut->defaultAction("test-daily", 1);
        $this->assertEquals("text/html", $response->contentType());
        Approvals::verifyHtml($response->output());
    }

    public function testListActionRendersListWithoutEntries(): void
    {
        $response = $this->sut->listAction("test-daily", 1);
        Approvals::verifyHtml($response->output());
    }

    public function testListActionRendersListWithAnEntry(): void
    {
        $this->listService->method('getDailyList')->willReturn([
            (object) ['range' => "9.", 'state' => "2", 'label' => "available"],
        ]);
        $response = $this->sut->listAction("test-daily", 1);
        Approvals::verifyHtml($response->output());
    }

    public function testListActionHandlesAjaxRequest(): void
    {
        $_SERVER['HTTP_X_REQUESTED_WITH'] = 'XMLHttpRequest';
        $_GET = ['ocal_name' => "test-daily"];
        $response = $this->sut->listAction("test-daily", 1);
        $this->assertEquals("text/html", $response->contentType());
        Approvals::verifyHtml($response->output());
    }

    public function testSaveActionReturnsEmptyResponseNameIsMissing(): void
    {
        $response = $this->sut->saveAction("test-daily", 1);
        $this->assertEquals("", $response->output());
    }

    public function testSaveActionReportsSuccess(): void
    {
        $_GET = ['ocal_name' => "test-daily"];
        $_POST = ['ocal_states' => json_encode(['2023-02' => array_fill(0, 27, "1")])];
        $response = $this->sut->saveAction("test-daily", 1);
        $this->assertEquals('<p class="xh_success">Successfully saved.</p>', $response->output());
    }
}
