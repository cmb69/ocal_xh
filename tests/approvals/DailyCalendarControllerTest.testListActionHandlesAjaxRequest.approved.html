
<script type="module" src="/ocal.js?&amp;v=1.4"></script>
<div class="ocal_lists" data-name="test-daily" data-ocal-config='{"message_unsaved_changes":"There are unsaved changes that will be lost when leaving the page!","isAdmin":false}'>
  
<p class="ocal_mode">
  <a class="ocal_button" href="/?&amp;ocal_action=calendar&amp;ocal_name=test-daily">Calendar view</a>
</p>
  
<div class="ocal_loaderbar"><img src="./images/ajax-loader-bar.gif" alt="loading"></div>
<div class="ocal_statusbar"></div>
  
<dl class="ocal_list">
  <dt>July 2023</dt>
  <dd class="ocal_noentry">There is no entry for this period.</dd>
</dl>
  
<p class="ocal_pagination">
  <a class="ocal_button" href="/?&amp;ocal_action=list&amp;ocal_name=test-daily&amp;ocal_year=2023&amp;ocal_month=7">Today</a>
  <a class="ocal_button" href="/?&amp;ocal_action=list&amp;ocal_name=test-daily&amp;ocal_year=2023&amp;ocal_month=8">Next Month</a>
  <a class="ocal_button" href="/?&amp;ocal_action=list&amp;ocal_name=test-daily&amp;ocal_year=2024&amp;ocal_month=7">Next Year</a>
</p>
</div>
